File: app/src/main/java/at/shockbytes/dante/util/books/RealmBookManager.java
package at.shockbytes.dante.util.books;

import android.support.annotation.NonNull;
import android.support.v4.app.FragmentActivity;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import at.shockbytes.dante.network.BookDownloader;
import at.shockbytes.dante.util.AppParams;
import at.shockbytes.dante.util.backup.BackupException;
import at.shockbytes.dante.util.backup.BackupManager;
import io.realm.Realm;
import io.realm.RealmResults;
import io.realm.Sort;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Func0;
import rx.schedulers.Schedulers;

/**
 * @author Martin Macheiner
 *         Date: 27.08.2016.
 */
public class RealmBookManager implements BookManager {

    private Realm realm;

    private BookDownloader bookDownloader;

    @Inject
    public RealmBookManager(BookDownloader bookDownloader, Realm realm) {
        this.realm = realm;
        this.bookDownloader = bookDownloader;
    }

    @Override
    public Book addBook(@NonNull Book book) {

        realm.beginTransaction();

        long id = getLastId();
        book.setId(id);
        realm.copyToRealm(book);

        realm.commitTransaction();

        return book;
    }

    @Override
    public Book getBook(long id) {
        return realm.where(Book.class).equalTo("id", id).findFirst();
    }

    @Override
    public void updateBookState(@NonNull final Book book, final Book.State newState) {

        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(@NonNull Realm realm) {

                book.setState(newState);
                realm.copyToRealmOrUpdate(book);
            }
        });

    }

    @Override
    public void updateCurrentBookPage(@NonNull final Book book, final int page) {

        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(@NonNull Realm realm) {

                book.setCurrentPage(page);
                realm.copyToRealmOrUpdate(book);
            }
        });

    }

    @Override
    public void updateBookStateAndPage(@NonNull final Book book, final Book.State state,
                                       final int page) {

        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(@NonNull Realm realm) {

                book.setCurrentPage(page);
                book.setState(state);
                realm.copyToRealmOrUpdate(book);
            }
        });
    }

    @Override
    public void removeBook(final long id) {

        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(@NonNull Realm realm) {
                realm.where(Book.class).equalTo("id", id).findFirst().deleteFromRealm();
            }
        });
    }

    @Override
    public Map<String, Integer> getStatistics() {

        Map<String, Integer> stats = new HashMap<>();

        int upcoming = realm.where(Book.class)
                .equalTo("ordinalState", Book.State.READ_LATER.ordinal()).findAll().size();
        int current = realm.where(Book.class)
                .equalTo("ordinalState", Book.State.READING.ordinal()).findAll().size();
        List<Book> doneList = realm.where(Book.class)
                .equalTo("ordinalState", Book.State.READ.ordinal()).findAll();
        int done = doneList.size();
        int pages = 0;
        for (Book b : doneList) {
            pages += b.getPageCount();
        }
        stats.put(AppParams.STAT_UPCOMING, upcoming);
        stats.put(AppParams.STAT_CURRENT, current);
        stats.put(AppParams.STAT_DONE, done);
        stats.put(AppParams.STAT_PAGES, pages);
        return stats;
    }

    @Override
    public Observable<BookSuggestion> downloadBook(@NonNull String isbn) {
        return bookDownloader.downloadBookSuggestion(isbn)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    @Override
    public Observable<List<Book>> getAllBooks() {
        List<Book> books = realm.where(Book.class)
                .findAll()
                .sort("id", Sort.DESCENDING);
        return Observable.just(books)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    @Override
    public Observable<Void> restoreBackup(FragmentActivity activity, List<Book> backupBooks,
                                          BackupManager.RestoreStrategy strategy) {

        switch (strategy) {

            case MERGE:
                return mergeBackupRestore(activity, backupBooks);

            case OVERWRITE:
                return overwriteBackupRestore(activity, backupBooks);

            default:
                return Observable.error(new BackupException("Cannot resolve restore strategy"));
        }
    }

    @Override
    public List<Book> getAllBooksSync() {
        return realm.where(Book.class)
                .findAll()
                .sort("id", Sort.DESCENDING);
    }

    @Override
    public Observable<List<Book>> getBookByState(Book.State state) {
        List<Book> books = realm.where(Book.class)
                .equalTo("ordinalState", state.ordinal())
                .findAll()
                .sort("id", Sort.DESCENDING);
        return Observable.just(books)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread());
    }

    @Override
    public void close() {
        realm.close();
    }

    /**
     * This must always be called inside a transaction
     */
    private long getLastId() {

        // This is the only place where a real realm transaction is executed
        BookConfig config = realm.where(BookConfig.class).findFirst();
        if (config == null) { // First time config is written
            config = realm.createObject(BookConfig.class);
        }

        return config.getLastPrimaryKey();
    }

    private Observable<Void> mergeBackupRestore(final FragmentActivity activity,
                                                final List<Book> backupBooks) {

        return Observable.defer(new Func0<Observable<Void>>() {
            @Override
            public Observable<Void> call() {

                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        List<Book> books = getAllBooksSync();
                        for (Book bBook : backupBooks) {

                            boolean insert = true;
                            for (Book sBook : books) {
                                if (sBook.getTitle().equals(bBook.getTitle())) {
                                    insert = false;
                                    break;
                                }
                            }

                            if (insert) {
                                addBook(bBook);
                            }
                        }
                    }
                });
                return Observable.empty();
            }
        });

    }

    private Observable<Void> overwriteBackupRestore(final FragmentActivity activity,
                                                    final List<Book> backupBooks) {

        return Observable.defer(new Func0<Observable<Void>>() {
            @Override
            public Observable<Void> call() {

                activity.runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        RealmResults<Book> stored = realm.where(Book.class).findAll();
                        realm.beginTransaction();
                        stored.deleteAllFromRealm();
                        realm.commitTransaction();

                        for (Book b : backupBooks) {
                            addBook(b);
                        }

                    }
                });


                return Observable.empty();
            }
        });



    }

}

File: app/src/main/java/at/shockbytes/dante/util/books/RealmBookManager.kt
package at.shockbytes.dante.util.books

import at.shockbytes.dante.backup.BackupManager
import at.shockbytes.dante.network.BookDownloader
import at.shockbytes.dante.util.AppParams
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.realm.Realm
import io.realm.Sort
import java.util.*


/**
 * @author Martin Macheiner
 * Date: 27.08.2016.
 */
class RealmBookManager(private val bookDownloader: BookDownloader,
                       private val realm: Realm) : BookManager {


    override val statistics: Map<String, Int>
        get() {
            val stats = HashMap<String, Int>()

            val upcoming = realm.where(Book::class.java)
                    .equalTo("ordinalState", Book.State.READ_LATER.ordinal).findAll().size
            val current = realm.where(Book::class.java)
                    .equalTo("ordinalState", Book.State.READING.ordinal).findAll().size
            val doneList = realm.where(Book::class.java)
                    .equalTo("ordinalState", Book.State.READ.ordinal).findAll()
            val done = doneList.size
            val pages = doneList.sumBy { it.pageCount }
            stats.put(AppParams.STAT_UPCOMING, upcoming)
            stats.put(AppParams.STAT_CURRENT, current)
            stats.put(AppParams.STAT_DONE, done)
            stats.put(AppParams.STAT_PAGES, pages)
            return stats
        }

    override val allBooks: Observable<List<Book>>
        get() {
            val books = realm.where(Book::class.java)
                    .findAll()
                    .sort("id", Sort.DESCENDING)
            return Observable.just<List<Book>>(books)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }

    override val allBooksSync: List<Book>
        get() = realm.where(Book::class.java).findAll().sort("id", Sort.DESCENDING)

    /**
     * This must always be called inside a transaction
     */
    private val lastId: Long
        get() {
            var config = realm.where(BookConfig::class.java).findFirst()
            if (config == null) {
                config = realm.createObject(BookConfig::class.java)
            }

            return config!!.lastPrimaryKey
        }

    override fun addBook(book: Book): Book {

        realm.beginTransaction()

        val id = lastId
        book.id = id
        realm.copyToRealm(book)

        realm.commitTransaction()

        return book
    }

    override fun getBook(id: Long): Book {
        return realm.where(Book::class.java).equalTo("id", id).findFirst()!!
    }

    override fun updateBookState(book: Book, newState: Book.State) {

        realm.executeTransaction { realm ->
            book.state = newState
            realm.copyToRealmOrUpdate(book)
        }

    }

    override fun updateCurrentBookPage(book: Book, page: Int) {

        realm.executeTransaction { realm ->
            book.currentPage = page
            realm.copyToRealmOrUpdate(book)
        }

    }

    override fun updateBookStateAndPage(book: Book, state: Book.State,
                                        page: Int) {

        realm.executeTransaction { realm ->
            book.currentPage = page
            book.state = state
            realm.copyToRealmOrUpdate(book)
        }
    }

    override fun removeBook(id: Long) {

        realm.executeTransaction { realm ->
            realm.where(Book::class.java)
                    .equalTo("id", id).findFirst()!!.deleteFromRealm()
        }
    }

    override fun downloadBook(isbn: String?): Observable<BookSuggestion> {
        return if (isbn == null) {
            Observable.error(NullPointerException("ISBN is null"))
        } else {
            bookDownloader.downloadBookSuggestion(isbn)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }
    }

    override fun restoreBackup(backupBooks: List<Book>, strategy: BackupManager.RestoreStrategy) {

        when (strategy) {

            BackupManager.RestoreStrategy.MERGE -> mergeBackupRestore(backupBooks)
            BackupManager.RestoreStrategy.OVERWRITE -> overwriteBackupRestore(backupBooks)
        }
    }

    override fun getBooksByState(state: Book.State): Observable<List<Book>> {
        val books = realm.where(Book::class.java)
                .equalTo("ordinalState", state.ordinal)
                .findAll()
                .sort("id", Sort.DESCENDING)
        return Observable.just<List<Book>>(books)
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
    }

    override fun close() {
        realm.close()
    }

    private fun mergeBackupRestore(backupBooks: List<Book>) {

        val books = allBooksSync
        for (bBook in backupBooks) {
            val insert = books.none { it.title == bBook.title }
            if (insert) {
                addBook(bBook)
            }
        }
    }

    private fun overwriteBackupRestore(backupBooks: List<Book>) {

        val stored = realm.where(Book::class.java).findAll()
        realm.beginTransaction()
        stored.deleteAllFromRealm()
        realm.commitTransaction()

        for (b in backupBooks) {
            addBook(b)
        }
    }

}
